src/gt4py/analysis/passes.py

```diff
@@ -1210,6 +1210,95 @@ def apply(self, transform_data: TransformData) -> TransformData:
+            self._reduce_fields()
+            return result
+
+        def _reduce_fields(self):
+            for field_name in self.reduced_fields:
+                field_decl = self.iir.fields[field_name]
+                field_decl.axes.pop()
+                for node in self.reduced_fields[field_name]["nodes"]:
+                    new_offset = {axis: node.offset[axis] for axis in field_decl.axes}
+                    node.offset = new_offset
+
+        def visit_StencilImplementation(
+            self, path: tuple, node_name: str, node: gt_ir.StencilImplementation
+        ) -> gt_ir.StencilImplementation:
+            self.iir = node
+            self.fields_extents = node.fields_extents.copy()
+            return self.generic_visit(path, node_name, node)
+
+        def visit_MultiStage(
+            self, path: tuple, node_name: str, node: gt_ir.MultiStage
+        ) -> Tuple[bool, Optional[gt_ir.MultiStage]]:
+            self.iteration_order = node.iteration_order
+            self.generic_visit(path, node_name, node)
+            return True, node
+
+        def visit_ApplyBlock(
+            self, path: tuple, node_name: str, node: gt_ir.ApplyBlock
+        ) -> Tuple[bool, gt_ir.ApplyBlock]:
+            self.interval = node.interval
+            self.generic_visit(path, node_name, node)
+            return True, node
+
+        def visit_FieldAccessor(
+            self, path: tuple, node_name: str, node: gt_ir.FieldAccessor
+        ) -> Tuple[bool, Optional[gt_ir.FieldAccessor]]:
+            self.fields_extents[node.symbol] |= node.extent
+            return True, node
+
+        def visit_FieldRef(
+            self, path: tuple, node_name: str, node: gt_ir.FieldRef
+        ) -> Tuple[bool, gt_ir.FieldRef]:
+            field_name = node.name
+            is_full_field = False
+            if field_name in self.iir.temporary_fields:
+                if self.iteration_order == gt_ir.IterationOrder.PARALLEL:
+                    is_full_field = True
+
+                elif field_name not in self.full_fields:
+                    extent = self.fields_extents[field_name]
+                    ndims = extent.ndims - 1
+                    if extent.lower_indices[ndims] == 0 and extent.upper_indices[ndims] == 0:
+                        if field_name not in self.reduced_fields:
+                            self.reduced_fields[field_name] = dict(
+                                interval=self.interval,
+                                nodes=list(),
+                            )
+                        elif self.interval != self.reduced_fields[field_name]["interval"]:
+                            is_full_field = True
+                    else:
+                        is_full_field = True
+
+                if field_name in self.reduced_fields:
+                    if is_full_field:
+                        self.full_fields.add(field_name)
+                        del self.reduced_fields[field_name]
```

Comment:
Nothing is ever added to `full_fields` unless it was considered reduced before. In general, I think it would be better to only maintain one of the lists (`full_fields` or `reduced_fields`,  that way, there is less potential for errors like this